teleop:
  ros__parameters:
#    update_rate: 100.0
    min_update_rate: 2.0

    # Set this true to display all the inputs. Very useful when trying to make a config!
    log_inputs: true

    input_sources:
      # Declare the names of your input sources here
      names: [
        "joy_input_source"
      ]
      joy_input_source:
        type: "teleop/JoyInputSource"

    control_modes:
      # Declare the names of your control modes here
      names: [
        "joint_space_control",
        "task_space_control"
      ]
      joint_space_control:
        type: "teleop/JointSpaceControlMode"
        controllers: [
          "nova_arm_velocity_controller"
        ]
      task_space_control:
        type: "teleop/TwistControlMode"
        controllers: [
          "nova_arm_position_controller",
          "nova_twistmapper"
        ]

    commands:
      # Declare the names of your commands here
      names: [
        "enter_twist_ik_control",
        "enter_joint_space_control",
        "log_locked",
        "log_unlocked"
      ]

      # Configure your commands here.
      enter_twist_ik_control:
        on: "twist_mode/down"
        type: "switch_control_mode"
        to: "twist_ik_control"
      enter_joint_space_control:
        on: "twist_mode/up"
        type: "switch_control_mode"
        to: "joint_space_control"

      # Locking and unlocking
      log_locked:
        on: "locked/down"
        type: "log"
        message: "Locked"
      log_unlocked:
        on: "locked/up"
        type: "log"
        message: "Unlocked"

joy_input_source:
  ros__parameters:
    topic: "/joy"
    axis_definitions: [
      "left_stick_x",
      "left_stick_y",
      "right_stick_x",
      "right_stick_y",
      "left_trigger",
      "right_trigger"
    ]
    button_definitions: [
      "cross",
      "circle",
      "square",
      "triangle",
      "share",
      "power",
      "start",
      "left_stick_click",
      "right_stick_click",
      "left_bumper",
      "right_bumper",
      "dpad_up",
      "dpad_down",
      "dpad_left",
      "dpad_right",
      "trackpad_click",
      "mute"
    ]

    # Convert input source names to meaningful mapped names
    # This logic is to be provided by the InputSourceManager when adding InputDeclarations as definitions for input
    # objects. The motivation is to share complex configuration functionality I created for joy inputs across any
    # inputs. This makes it easier to implement other input sources in the future and reuse code in a way that allows
    # all input sources to benefit from potential optimizations made to remapping.
    #
    # TODO: Should modification be made to the inputs directly?
    remap:
      axes:
        J1:
          from: "left_stick_x"
        J2:
          from: "left_stick_y"
          invert: true
          # Remap the range of the input from [-1, 1] to [0, 1]
          range:
            in: [-1, 1]
            out: [0, 1]
            # Optionally, clamp the value
            limit: true
        J3:
          from: "left_stick_x"
          # Remap the range of the input from [-1, 1] to [0, 1]
          range:
            in: [-1.0, 1.0]
            # You can also clamp the value without remapping the output range
            limit: true
        J4:
          from: "right_stick_x"
          in: -1.0
          out_min: 0.0
        J5.from: "right_stick_y"
        J6:
          from: "dpad_x"
          from_buttons:
            positive: "dpad_right"
            negative: "dpad_left"

joint_space_control:
  ros__parameters:
    # TODO: get the actual name from arm code
    topic: "/todo_joint_space"

    joint_definitions: [
      "arm_j1",
      "arm_j2",
      "arm_j3",
      "arm_j4",
      "arm_j5",
      "arm_j6"
    ]

    joints:
      arm_j1:
        input_name: "J1"
        max_speed: 0.364467013889
      arm_j2:
        input_name: "J2"
        max_speed: 0.364467013889
      arm_j3:
        input_name: "J3"
        max_speed: 0.364467013889
      arm_j4:
        input_name: "J4"
        max_speed: 0.364467013889
      arm_j5:
        input_name: "J5"
        max_speed: 0.364467013889
      arm_j6:
        input_name: "J6"
        max_speed: 0.364467013889

task_space_control:
  ros__parameters:
    # TODO: get the actual name from arm code
    topic: "/todo_task_space"


# trigger:
#   id: 0
# top down:
#   id: 1
# top left:
#   id: 2
# top right:
#   id: 0
